File: CCVTAC/CCVTAC.Console/History.cs
using System.IO;
using System.Text.Json;
using Spectre.Console;

namespace CCVTAC.Console;

/// <summary>
/// Handles storing, retrieving, and (eventually) analyzing data relating
/// to URLs that the user has entered.
/// </summary>
public class History
{
    private static readonly char Separator = ';';
    private string FilePath { get; init; }

    public History(string filePath)
    {
        FilePath = filePath;
    }

    /// <summary>
    /// Add a URL and related data to the history file.
    /// </summary>
    public void Append(string url, DateTime entryTime, Printer printer)
    {
        try
        {
            string serializedEntryTime = JsonSerializer.Serialize(entryTime).Replace("\"", "");
            File.AppendAllText(FilePath, serializedEntryTime + Separator + url + Environment.NewLine);
            printer.Print($"Added \"{url}\" to the history log.");
        }
        catch (Exception ex)
        {
            printer.Error($"Could not append URL {url} to history log: " + ex.Message);
        }
    }

    public void PrintRecent(Printer printer)
    {
        try
        {
            IEnumerable<IGrouping<DateTime, string>> historyData =
                File.ReadAllLines(FilePath)
                    .TakeLast(25)
                    .Select(line => line.Split(Separator))
                    .Where(lineItems => lineItems.Length == 2)
                    .GroupBy(line => DateTime.Parse(line[0]), line => line[1]);

            Table table = new();
            table.AddColumns("Time", "URL");

            string formattedTime, urls;
            foreach (var thisDate in historyData)
            {
                formattedTime = $"{thisDate.Key:dd MMMM yyyy HH:mm:ss}";
                urls = string.Join(Environment.NewLine, thisDate);
                table.AddRow(formattedTime, urls);
            }

            printer.Print(table);
        }
        catch (Exception ex)
        {
            printer.Error($"Could not print recent history: {ex.Message}");
        }
    }
}

File: src/CCVTAC.Main/History.fs
namespace CCVTAC.Main

open CCVTAC.Main.IoUtilities.Files
open System
open System.IO
open System.Text.Json
open Spectre.Console

type History(filePath: string, displayCount: int) =

    let separator = ';'

    member private _.FilePath = filePath
    member private _.DisplayCount = displayCount

    /// Write a URL and its related data to the history file.
    member this.Append(url: string, entryTime: DateTime, printer: Printer) : unit =
        try
            let serializedTime = JsonSerializer.Serialize(entryTime).Replace("\"", "")
            let text = serializedTime + string separator + url + String.newLine

            match appendToFile this.FilePath text with
            | Ok _      -> printer.Debug $"Added \"%s{url}\" to the history log."
            | Error err -> printer.Error $"Failed to write \"%s{url}\" to the history log at \"{this.FilePath}\": {err}"
        with exn ->
            printer.Error $"Could not append URL(s) to history log: {exn.Message}"

    member this.ShowRecent(printer: Printer) : unit =
        try
            let lines =
                File.ReadAllLines this.FilePath
                |> Seq.rev
                |> Seq.truncate this.DisplayCount
                |> Seq.rev
                |> Seq.toList

            let historyData =
                lines
                |> Seq.choose (fun line ->
                    match line.Split separator with
                    | [| dateText; url |] ->
                        match DateTime.TryParse dateText with
                        | true, date -> Some (date, url)
                        | _ -> None
                    | _ -> None)
                |> Seq.groupBy fst
                |> Seq.map (fun (dt, pairs) -> dt, pairs |> Seq.map snd |> Seq.toList)

            // TODO: These presentation matters shouldn't be here.
            let table = Table()
            table.Border <- TableBorder.None
            table.AddColumns("Time", "URL") |> ignore
            table.Columns[0].PadRight(3) |> ignore

            for dateTime, urls in historyData do
                let formattedTime = sprintf "%s" (dateTime.ToString("yyyy-MM-dd HH:mm:ss"))
                let joinedUrls = String.Join(String.newLine, urls)
                table.AddRow(formattedTime, joinedUrls) |> ignore

            Printer.PrintTable table
        with exn ->
            printer.Error $"Could not display history: %s{exn.Message}"
